File: ZadorozhnyyAlexander/Semester 4/Block 1/Fibers/Fibers/Fiber.cs
using System;

namespace Fibers
{
    public class Fiber
    {

        /// <summary>
        /// The fiber action delegate.
        /// </summary>
        private Action action;

        /// <summary>
        /// Gets the fiber identifier.
        /// </summary>
        public uint Id { get; private set; }

        /// <summary>
        /// Gets the id of the primary fiber.
        /// </summary>
        /// <remarks>If the Id is 0 then this means that there is no primary Id on the fiber.</remarks>
        public static uint PrimaryId { get; private set; }

        /// <summary>
        /// Gets the flag identifing the primary fiber (a fiber that can run other fibers).
        /// </summary>
        public bool IsPrimary { get; private set; }

        /// <summary>
        /// Initializes a new instance of the <see cref="Fiber"/> class.
        /// </summary>
        /// <param name='action'>Action.</param>
        public Fiber(Action action)
        {
            InnerCreate(action);
        }

        /// <summary>
        /// Deletes the current fiber.
        /// </summary>
        /// <remarks>This method should only be used in the fiber action that's executing.</remarks>
        public void Delete()
        {
            UnmanagedFiberAPI.DeleteFiber(Id);
        }

        /// <summary>
        /// Deletes the fiber with the specified fiber id.
        /// </summary>
        /// <param name='fiberId'>fiber id.</param>
        public static void Delete(uint fiberId)
        {
            UnmanagedFiberAPI.DeleteFiber(fiberId);
        }

        /// <summary>
        /// Switches the execution context to the next fiber.
        /// </summary>
        /// <param name='fiberId'>Fiber id.</param>
        public static void Switch(uint fiberId)
        {
            // for debug only and to show that indeed it works! Remove this line!!!
            Console.WriteLine(string.Format("Fiber [{0}] Switch", fiberId));

            UnmanagedFiberAPI.SwitchToFiber(fiberId);
        }

        /// <summary>
        /// Creates the fiber.
        /// </summary>
        /// <remarks>This method is responsible for the *actual* fiber creation.</remarks>
        /// <param name='action'>Fiber action.</param>
        private void InnerCreate(Action action)
        {
            this.action = action;

            if (PrimaryId == 0)
            {
                PrimaryId = UnmanagedFiberAPI.ConvertThreadToFiber(0);
                IsPrimary = true;
            }

            UnmanagedFiberAPI.LPFIBER_START_ROUTINE lpFiber = FiberRunnerProc;
            Id = UnmanagedFiberAPI.CreateFiber(100500, lpFiber, 0);
        }

        /// <summary>
        /// Fiber method that executes the fiber action.
        /// </summary>
        /// <param name='lpParam'>Lp parameter.</param>
        /// <returns>fiber status code.</returns>
        private uint FiberRunnerProc(uint lpParam)
        {
            uint status = 0;

            try
            {
                action();
            }
            catch (Exception)
            {
                status = 1;
                throw;
            }
            finally
            {
                if (status == 1)
                    UnmanagedFiberAPI.DeleteFiber((uint)Id);
            }

            return status;
        }
    }
}

File: IgorKorobitsyn/Block1Task1/src/main/kotlin/Fiber.kt
package fibers

import java.lang.Exception

class Fiber(private var action: () -> Unit){

    private var id: Int = 0

    fun getId(): Int {
        return id
    }

    private var primaryId: Int = 0

    private var isPrimary: Boolean = false

    fun isPrimary(): Boolean{
        if (isPrimary && id == primaryId){
            isPrimary = true
            return true
        }
        else{
            isPrimary = false
            return false
        }
    }

    init {
        create()
    }

    private fun create(){
        if (primaryId == 0){
            primaryId = UnmanagedFiberAPI.count // = UnmanagedFiberAPI.kernel32.ConvertThreadToFiber(0)
            isPrimary = true;
        }
        val lpFiber = EventCallbackInterface { param -> fiberRunnerProc(param) }


        id = UnmanagedFiberAPI.count++ // = UnmanagedFiberAPI.kernel32.CreateFiber(10050, lpFiber, 0)
    }

    private fun fiberRunnerProc(param: Int): Int{
        var status: Int = 0
        try{
            action()
        } catch (e: Exception) {
            status = 1
            throw e
        } finally {
            if (status == 1){
                //UnmanagedFiberAPI.kernel32.DeleteFiber(id)
            }
        }
        return status;
    }

    fun delete(){
        delete(id)
    }


    fun delete(fiberId: Int){
        //UnmanagedFiberAPI.kernel32.DeleteFiber(fiberId)
    }

    fun switch(fiberId: Int){
        //UnmanagedFiberAPI.kernel32.SwitchToFiber(fiberId)
    }
}